src/crypto/ecdh.rs

```diff
@@ -19,87 +19,175 @@ const ANON_SENDER: [u8; 20] = [
     0x20, 0x20, 0x20, 0x20,
 ];
 
-const SECRET_KEY_LENGTH: usize = 32;
-
-/// Secret key for ECDH with Curve25519, the only combination we currently support.
+/// Secret key for ECDH
 #[derive(Clone, PartialEq, Eq, Zeroize, ZeroizeOnDrop)]
-pub struct SecretKey {
-    /// The secret point.
-    pub secret: [u8; 32],
-    pub hash: HashAlgorithm,
-    pub oid: Vec<u8>,
-    pub alg_sym: SymmetricKeyAlgorithm,
+pub enum SecretKey {
+    // ECDH with Curve25519
+    Curve25519Legacy {
+        /// The secret point.
+        secret: [u8; ECCCurve::Curve25519.secret_key_length()],
+        hash: HashAlgorithm,
+        alg_sym: SymmetricKeyAlgorithm,
+    },
+
+    // ECDH with Nist P256
+    P256 {
+        /// The secret point.
+        secret: [u8; ECCCurve::P256.secret_key_length()],
+        hash: HashAlgorithm,
+        alg_sym: SymmetricKeyAlgorithm,
+    },
+
+    // ECDH with Nist P384
+    P384 {
+        /// The secret point.
+        secret: [u8; ECCCurve::P384.secret_key_length()],
+        hash: HashAlgorithm,
+        alg_sym: SymmetricKeyAlgorithm,
+    },
+
+    // ECDH with Nist P521
+    P521 {
+        /// The secret point.
+        secret: [u8; ECCCurve::P521.secret_key_length()],
+        hash: HashAlgorithm,
+        alg_sym: SymmetricKeyAlgorithm,
+    },
 }
 
 impl fmt::Debug for SecretKey {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let (curve, alg_sym, hash) = self.key_params();
+
         f.debug_struct("EcdhSecretKey")
             .field("secret", &"[..]")
-            .field("hash", &self.hash)
-            .field("oid", &hex::encode(&self.oid))
-            .field("alg_sym", &self.alg_sym)
+            .field("hash", &hash)
+            .field("oid", &hex::encode(curve.oid()))
+            .field("alg_sym", &alg_sym)
             .finish()
     }
 }
 
 impl KeyParams for SecretKey {
-    type KeyParams = (Vec<u8>, SymmetricKeyAlgorithm, HashAlgorithm);
+    type KeyParams = (ECCCurve, SymmetricKeyAlgorithm, HashAlgorithm);
 
     fn key_params(&self) -> Self::KeyParams {
-        (self.oid.clone(), self.alg_sym, self.hash)
+        match self {
+            SecretKey::Curve25519Legacy { hash, alg_sym, .. } => {
+                (ECCCurve::Curve25519, *alg_sym, *hash)
+            }
+            SecretKey::P256 { hash, alg_sym, .. } => (ECCCurve::P256, *alg_sym, *hash),
+            SecretKey::P384 { hash, alg_sym, .. } => (ECCCurve::P384, *alg_sym, *hash),
+            SecretKey::P521 { hash, alg_sym, .. } => (ECCCurve::P521, *alg_sym, *hash),
+        }
     }
 }
 
 impl Decryptor for SecretKey {
     fn decrypt(&self, mpis: &[Mpi], fingerprint: &[u8]) -> Result<Vec<u8>> {
+        /// Generic for Rust Crypto curves
+        fn derive_shared_secret<C>(
```

Comment:
lets put this function outside of the function, not a big fan of nesting function defs, other for very short helpers